api/src/test/java/org/apache/iceberg/TestSchema.java

```diff
@@ -0,0 +1,111 @@
+              Types.StructType.of(
+                  Types.NestedField.optional(6, "inner_ts", Types.TimestampNanoType.withZone()),
+                  Types.NestedField.required(7, "data", Types.StringType.get()))),
+          Types.NestedField.optional(
+              8,
+              "struct_arr",
+              Types.StructType.of(
+                  Types.NestedField.optional(9, "ts", Types.TimestampNanoType.withoutZone()))));
+
+  private static final Schema INITIAL_DEFAULT_SCHEMA =
+      new Schema(
+          Types.NestedField.required(1, "id", Types.LongType.get()),
+          Types.NestedField.required("has_default")
+              .withId(2)
+              .ofType(Types.StringType.get())
+              .withInitialDefault("--")
+              .withWriteDefault("--")
+              .build());
+
+  private static final Schema WRITE_DEFAULT_SCHEMA =
+      new Schema(
+          Types.NestedField.required(1, "id", Types.LongType.get()),
+          Types.NestedField.required("has_default")
+              .withId(2)
+              .ofType(Types.StringType.get())
+              .withWriteDefault("--")
+              .build());
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 2})
+  public void testUnsupportedTimestampNano(int formatVersion) {
+    Assertions.assertThatThrownBy(() -> Schema.checkCompatibility(TS_NANO_CASES, formatVersion))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage(
+            "Invalid schema for v%s:\n"
+                + "- Invalid type for ts: timestamptz_ns is not supported until v3\n"
+                + "- Invalid type for arr.element: timestamp_ns is not supported until v3\n"
+                + "- Invalid type for struct.inner_ts: timestamptz_ns is not supported until v3\n"
+                + "- Invalid type for struct_arr.ts: timestamp_ns is not supported until v3",
+            formatVersion);
+  }
+
+  @Test
+  public void testSupportedTimestampNano() {
+    Assertions.assertThatCode(() -> Schema.checkCompatibility(TS_NANO_CASES, 3))
+        .doesNotThrowAnyException();
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 2})
+  public void testUnsupportedInitialDefault(int formatVersion) {
+    Assertions.assertThatThrownBy(
+            () -> Schema.checkCompatibility(INITIAL_DEFAULT_SCHEMA, formatVersion))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage(
+            "Invalid schema for v%s:\n"
+                + "- Invalid initial default for has_default: "
+                + "non-null default (--) is not supported until v3",
+            formatVersion);
+  }
+
+  @Test
+  public void testSupportedInitialDefault() {
+    Assertions.assertThatCode(() -> Schema.checkCompatibility(INITIAL_DEFAULT_SCHEMA, 3))
+        .doesNotThrowAnyException();
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 2, 3})
+  public void testSupportedInitialDefault(int formatVersion) {
+    // only the initial default is a forward-incompatible change
```

Comment:
nit: We could make this self contained to the test by saying 

Writer defaults are universally compatible

Or something?